tests/task.py

```diff
@@ -0,0 +1,123 @@
+from os import getenv as os_getenv
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+import torch.optim as optim
+from torch.utils.data import DataLoader, Subset
+
+from .model import LinearMLP
+from .dataset import XORDataset, ExtraXORDataset
+
+
+def use_cuda():
+    if int(os_getenv('CPU_ONLY', '0')):
+        return False
+    else:
+        return torch.cuda.is_available()
+
+
+class TaskTemplate(type):
+    def __call__(cls, *args, **kwargs):
+        obj = type.__call__(cls, *args, **kwargs)
+        if hasattr(obj, '__post_init__'):
+            obj.__post_init__()
+        return obj
+
+
+def use_metaclass(meta_cls):
+    """ A decorator to make metaclass work on both Py2k and Py3k. """
+    def wrapper(cls):
+        attrs = vars(cls).copy()
+        attrs.pop('__dict__', None)
+        attrs.pop('__weakref__', None)
+        return meta_cls(cls.__name__, cls.__bases__, attrs)
+    return wrapper
+
+
+@use_metaclass(TaskTemplate)
+class BaseTask(object):
+    def __init__(self):
+        self.batch_size = -1
+        self.model = None
+        self.optimizer = None
+        self.criterion = None
+        self.device = None
+        self.train_loader = None
+        self.val_loader = None
+
+    def __post_init__(self):
+        # check whether cuda is available or not
+        if not use_cuda():
+            self.device = None
+        if self.device is None:
+            return
+
+        if isinstance(self.device, str):
+            try:
+                self.device = torch.device(self.device)
+            except RuntimeError as ex:
+                raise ex
```

Comment:
Oh! I was trying to validate the device string by calling `torch.device()` and re-raise a `ValueError` when given device string is not supported. Like this:
```python
try:
    self.device = torch.device(self.device)
except RuntimeError as ex:
    raise ValueError('Invalid device string.') from ex
```
By doing this, we don't need to maintain a list for supported devices here.

But since I found that syntax of `raise ex2 from ex1` isn't supported in Py2k, I rewrote that line.
However, it seems the try-except block is redundant here. I'll remove the try-except block in the following commits.

Reply:
Since we are going to remove support of Py2k, I think it is still a good idea to remove this try-except block. Let's just raise that `RuntimeError` in a fail-fast style, and it would be a clear error message for developers like us.